cookbook/configuration/micro-kernel-trait.rst

```diff
@@ -0,0 +1,327 @@
+    $request = Request::createFromGlobals();
+    $response = $kernel->handle($request);
+    $response->send();
+    $kernel->terminate($request, $response);
+
+That's it! To test it, you can start the built-in web server:
+
+.. code-block:: bash
+
+    $ php -S localhost:8000
+
+Then see the JSON response in your browser:
+
+> http://localhost:8000/random/10
+
+The Methods of a "Micro" Kernel
+-------------------------------
+
+When you use the ``MicroKernelTrait``, your kernel needs to have exactly three methods
+that define your bundles, your services and your routes:
+
+**registerBundles()**
+    This is the same ``registerBundles()`` that you see in a normal kernel.
+
+**configureContainer(ContainerBuilder $c, LoaderInterface $loader)**
+    This methods builds and configures the container. In practice, you will use
+    ``loadFromExtension`` to configure different bundles (this is the equivalent)
+    of what you see in a normal ``config.yml`` file. You can also register services
+    directly in PHP or load external configuration files (shown below).
+
+**configureRoutes(RouteCollectionBuilder $routes)**
+    Your job in this method is to add routes to the system. The ``RouteCollectionBuilder``
+    is new in Symfony 2.8 and has methods that make adding routes in PHP more fun.
+    You can also load external routing files (shown below).
+
+
+Advanced Example: Twig, Annotations and the Web Debug Toolbar
+-------------------------------------------------------------
+
+The purpose of the ``MicroKernelTrait`` is *not* to have a single-file application.
+Instead, it's goal to give you the power to choose your bundles and structure.
+
+First, you'll probably want to put your PHP classes in an ``src/`` directory. Configure
+your ``composer.json`` file to load from there:
+
+... code-block:: json
+
+{
+    "require": {
+        "...": "...",
+    },
+    "autoload": {
+        "psr-4": {
+            "": "src/"
+        }
+    }
+}
+
+Now, suppose you want to use Twig and load routes via annotations. For annotation
+routing, you need SensioFrameworkExtraBundle. This comes with a normal Symfony project.
+But in this case, you need to download it:
+
+.. code-block:: bash
+
+    composer require sensio/framework-extra-bundle
+
+Instead of putting *everything* in ``index.php``, create a new ``app/AppKernel.php``
+to hold the kernel. Now it looks like this::
+
+    // app/AppKernel.php
+
+    use Symfony\Bundle\FrameworkBundle\Kernel\MicroKernelTrait;
+    use Symfony\Component\Config\Loader\LoaderInterface;
+    use Symfony\Component\DependencyInjection\ContainerBuilder;
+    use Symfony\Component\HttpKernel\Kernel;
+    use Symfony\Component\Routing\RouteCollectionBuilder;
+    use Doctrine\Common\Annotations\AnnotationRegistry;
+
+    // require Composer's autoloader
+    $loader = require __DIR__.'/../vendor/autoload.php';
+    // auto-load annotations
+    AnnotationRegistry::registerLoader(array($loader, 'loadClass'));
```

Comment:
There is some weird behavior here. Yes, you do need this - I just double-checked by removing that code (and clearing the cache) and I got the error:

>
[Semantical Error] The annotation "@Sensio\Bundle\FrameworkExtraBundle\Configuration\Route"
> in method App\Controller\MicroController::randomAction() does not exist, or could not be auto-loaded.

But, the other day in a different project, I *swear* I removed this, cleared the cache, and *didn't* get that error. For now, I'll include it - but yes, I'd love for this to happen automatically.